testing/src/scenario/context.py

```diff
@@ -736,6 +737,12 @@ def __init__(
 
         self.on = CharmEvents()
 
+    @property
+    def charm_spec(self):
+        """Deprecated property for accessing the Context's charm spec."""
+        warnings.warn('Accessing Context.charm_spec is deprecated.', DeprecationWarning)
```

Comment:
It's nice to provide alternatives in these sort of things. Tricky when we don't have a suggestion, which perhaps indicates that it's not ideal to deprecate.

Maybe we could suggest just for the metadata, like "... is deprecated. To generate a State based on the metadata, consider using State.from_context().". And assume people will figure something out for the other attributes.

Reply:
Aside: I suspect that `stacklevel=2,` is needed here.

Reply:
It would be nice to make this surface clearly in people's IDEs as deprecated (I'm assuming emitting a deprecation warning at runtime doesn't do this). Either with a fancy type annotation or some other annotation that IDEs understand.